### ISSUE TITLE
Add support for nested arrays

### ISSUE THREAD

**Issue by rage-shadowman** [OAI/OpenAPI-Specification#40]
Some industry standards (such as GeoJSON) use nested arrays in their JSON models.

See:
Spec in English: http://www.geojson.org/geojson-spec.html
Spec in JSON: https://github.com/fge/sample-json-schemas/blob/master/geojson/geometry.json

A Geometry object has "type" property of "Point", "LineString", "Polygon", "MultiPoint", "MultiLineString", "MultiPolygon" or "GeometryCollection". It also has optional "coordinates" property (required for all values of "type" except "GeometryCollection") which would be an array of something (depending on the value of "type"). The other properties aren't relevant to this request.

A Point is a Geometry object with a "type" property of value "Point" and a "coordinates" property of an array of 2 or 3 doubles (lon/lat coordinates and optional elevation). So this one you can already handle (except for the min/max item count) as array[double].

A LineString is a Geometry object with a "type" property of value "LineString" and a "coordinates" property of an array of 2 or more connected Point coordinate arrays (so an array[array[double]]).

A MultiLineString is a Geometry object with a "type" property of value "MultiLineString" and a "coordinates" property of an array of 1 or more LineString coordinate arrays (so an array[array[array[double]]]).

We have other use cases for them but this particular case is an industry standard, so I thought it best to use that example here.

See also:
https://github.com/wordnik/swagger-core/issues/444
https://github.com/wordnik/swagger-core/issues/449

**Comment by newmanw** [comment]
Here is a good JSON spec for GeoJSON:
https://github.com/fge/sample-json-schemas/tree/master/geojson

Not sure how to put together a swagger spec to define the same thing without 'oneOf'. That being said, I am a serious swagger rookie so it could very well be possible. 

Has anyone tried to put together a swagger spec to outline something like a FeatureCollection?

**Comment by webron** [comment]
From the looks of it, FeatureCollection *can* be described with Swagger, using inheritance. Since all Geometry objects contain a `type` field, the root object (called Geometry for example) could contain that property and used as a discriminator. The other objects would use `allOf` to include it and add their own properties.

**Comment by newmanw** [comment]
Sounds good, I will give it a go and post. A quick question how would I represent a FeatureCollection response, an array of any of the geometry types?

**Comment by webron** [comment]
No. It looks like FeatureCollection is an object with two properties: `type` and 'features'. The `features' value is an array of the geometry types.

**Comment by newmanw** [comment]
Yes, you are correct, sorry for the confusion.  What I really meant was how to represent the features array.

Assuming I had ssd inheritance to define some of the geometry types (pointFeature, polyFeature):

definitions:
  ...
  baseFeature:
    # define base feature here in which concrete features can extend
  pointFeature:
    # extend from base feature, add point array
  polyFeature:
    # extend from base feature add array of point arrays
  feature:
    required:
       - geometry
       - properties
    properties:
      type:
           type: 'string'
      geometry:
        oneOf:
           - $ref: '#/definitions/pointFeature
           - $ref: '#/definitions/polyFeature
           # etc...
  features:
    properties:
      type:
        type: array
        items: 
          $ref: '#/definitions/feature

Where this breaks is obviously the use of oneOf.  Confusion on my part is how to represent an array ('features') in which items in that array can be any of the concrete geometry objects.

Hopefully I am not confusing the issue more :(

**Comment by webron** [comment]
You won't be using oneOf at all. Since inheritance is in place, anything that extends the Geometry object (that is, uses allOf to include the Geometry object) could be used in place where the Geometry object is used.

**Comment by newmanw** [comment]
Ah ok, so just use the 'abstract' geometry as the reference for the geometry property in a feature.

Like so?

```yaml
definitions:
  ...
  geometry:
    required:
      - type
    properties:
      type:
        type: string # should probably be enum, but string ok for this example
  point:
    allOf:
      - $ref: '#/definitions/geometry
      - coordinates: 
        type: array
        items:
          type: double
  polygon:
    # TODO define polygon, extend geometry
 feature:
    required:
       - geometry
       - properties
    properties:
      type:
           type: 'string'
      geometry:
         $ref: '#/definitions/geometry'
  features:
    properties:
      type:
        type: array
        items: 
          $ref: '#/definitions/feature'
```

Hopefully its obvious that this is not a complete schema and it not validated ;)

I will try to write up the definitions and post in case anyone else is interested.

Thanks for the help @webron

**Comment by webron** [comment]
Yeah, something along those lines, though you should really use the `discriminator` property as well (https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#schemaDiscriminator).

Keep in mind it's quite possible that neither the editor nor the ui support inheritance yet (I think there's in open ticket on the ui, not sure about the editor).

**Comment by newmanw** [comment]
+1 on discriminator, thanks for the tip!

**Comment by webron** [comment]
Keep in mind that without the discriminator it's not inheritance, only composition.

**Comment by alotau** [comment]
@newmanw If you managed to complete any of the definitions for GeoJSON, it would be awesome if you could share.  I'm wresting with swagger now trying to geometry/point/linestring stuff in there.  Thanks in advance.

**Comment by derekperkins** [comment]
I'd love to see it too.

**Comment by logantpowell** [comment]
:+1:

**Comment by FLASHHAMMER** [comment]
I'm having a similar issue. I created a server that uses Apache Axis 1.2.1 to autogenerate the classes from a pretty big WSDL SOAP service that is the core project of the business back-end, that have multiple endpoints. Most of the exposed methods/APIs have multiple nested custom object type arrays in the requests and responses. Tried using Swagger and found out pretty quick this limitation with arrays of custom object type arrays. I am using the Swagger 3 auto configuration for Spring Boot. I am having more issues with reflection, but the Swagger limitation in this matter really made me move away from Swagger and OpenAPI.
